### PR TITLE
ENHANCEMENT Better optimised class name migration with substantially fewer queries

### LINKED ISSUE #9752: Design of classname_value_remapping results in massively slow dev/build
## Affected Version

SilverStripe 4.x (tested with framework 4.5.3)

## Description

When running a dev/build legacy classnames will be remapped. However with a large enough project with a large enough number of classnames, the performance of this drops exponentially.

On one specific project there was a total of 67 classname_value_remapping and 142 dataobjects. This resulted in almost 10,000 queries to simply check if there were any rows matching (deployments were timing out, so I don't know how long it would take to run).

## Steps to Reproduce

Install a project with a huge number of dataobjects and classname mappings. Try to do a dev/build. :)

## Suggestion

Debugging showed that the queries being run were checking base classes for objects which could not possibly have had the queried legacy values. I pasted a short sample of these queries into the below gist.

https://gist.github.com/tractorcow/3d9d8084fe3cdf1cb6253677ad73fee2

My suggestion is that loop is optimised to only call updateLogacyClassnames() on classes which match, comparing $className to $newClassName, instead of all classes.

```php
            // Remap obsolete class names
            $remappingConfig = $this->config()->get('classname_value_remapping');
            $remappingFields = $this->getClassNameRemappingFields();

            foreach ($remappingFields as $className => $fieldNames) {
                foreach ($fieldNames as $fieldName) {
                    foreach ($remappingConfig as $oldClassName => $newClassName) {
                        $this->updateLegacyClassNames($className, $fieldName, $oldClassName, $newClassName);
                    }
                }
            }
```

For example, the legacy ClassName value `PermissionRoleCode` should be checked against the `PermissionRoleCode` table, but not against the `WebConsole_BackgroundTask` table.

### PR DESCRIPTION
**Pull request by tractorcow**
Fixes #9752

Reduces complexity from o(n^2) to o(n)

Request review:
 - is this the right place to put the code, or should I split this out into a dev task in this change?
 - is 4.6 the right branch? Not sure if you want to treat this as a bugfix or api change.
 - is the approach acceptable? (replace the individual with a more complex case statement in assignment).

### CODE REVIEW

File: src/ORM/DatabaseAdmin.php [hunk 1]
```diff
@@ -412,8 +406,23 @@ public function doBuild($quiet = false, $populate = true, $testMode = false)
      * @param string $fieldName    The field name to look in for obsolete class names
      * @param string $oldClassName The old class name
      * @param string $newClassName The new class name
+     * @deprecated This method performs inefficiently so it's been disabled
      */
     protected function updateLegacyClassNames($dataClass, $fieldName, $oldClassName, $newClassName)
+    {
+        // no-op
+        Deprecation::notice('5.0');
```
Comment (ScopeyNZ) [inline src/ORM/DatabaseAdmin.php:414]: Couldn't you leave the existing functionality intact by just adding

```suggestion
        Deprecation::notice('5.0');
        $this->updateLegacyClassNameField($dataClass, $fieldName, [$oldClassName => $newClassName]);
```

I think we should switch any core code to use the new API, but I don't really think we can just "disable" the functionality of a method in a patch version.

Comment (tractorcow) [reply src/ORM/DatabaseAdmin.php:414]: True :D

File: src/ORM/DatabaseAdmin.php [hunk 2]
```diff
@@ -412,8 +406,24 @@ public function doBuild($quiet = false, $populate = true, $testMode = false)
      * @param string $fieldName    The field name to look in for obsolete class names
      * @param string $oldClassName The old class name
      * @param string $newClassName The new class name
+     * @deprecated 5.0 use updateLegacyClassNameField instead
      */
     protected function updateLegacyClassNames($dataClass, $fieldName, $oldClassName, $newClassName)
+    {
+        // no-op
```
Comment (ScopeyNZ) [inline src/ORM/DatabaseAdmin.php:413]:
```suggestion
```

One last pedantic request then I'm happy 😄

**Review by ScopeyNZ — APPROVED** [review]

**Comment by tractorcow** [comment]
It is done :)

**Comment by ScopeyNZ** [comment]
I've left this a bit for any other feedback, but it's not getting any so let's just merge it.

**Pull request merged**